pilot/platform/kube/inject/inject.go

```diff
@@ -322,172 +330,51 @@ func timeString(dur *duration.Duration) string {
 	return out.String()
 }
 
-func injectIntoSpec(p *Params, spec *v1.PodSpec, metadata *metav1.ObjectMeta) {
-	// proxy initContainer 1.6 spec
-	initArgs := []string{
-		"-p", fmt.Sprintf("%d", p.Mesh.ProxyListenPort),
-		"-u", strconv.FormatInt(p.SidecarProxyUID, 10),
-	}
-	if p.IncludeIPRanges != "" {
-		initArgs = append(initArgs, "-i", p.IncludeIPRanges)
-	}
-
-	var pullPolicy v1.PullPolicy
-	switch p.ImagePullPolicy {
-	case "Always":
-		pullPolicy = v1.PullAlways
-	case "IfNotPresent":
-		pullPolicy = v1.PullIfNotPresent
-	case "Never":
-		pullPolicy = v1.PullNever
-	default:
-		pullPolicy = v1.PullIfNotPresent
-	}
-
-	privTrue := true
-
-	initContainer := v1.Container{
-		Name:            InitContainerName,
-		Image:           p.InitImage,
-		Args:            initArgs,
-		ImagePullPolicy: pullPolicy,
-		SecurityContext: &v1.SecurityContext{
-			Capabilities: &v1.Capabilities{
-				Add: []v1.Capability{"NET_ADMIN"},
-			},
-			// TODO: Determine SELINUX options needed to remove privileged
-			Privileged: &privTrue,
-		},
-	}
-
-	enableCoreDumpContainer := v1.Container{
-		Name:    enableCoreDumpContainerName,
-		Image:   enableCoreDumpImage,
-		Command: []string{"/bin/sh"},
-		Args: []string{
-			"-c",
-			fmt.Sprintf("sysctl -w kernel.core_pattern=%s/core.%%e.%%p.%%t && ulimit -c unlimited",
-				p.Mesh.DefaultConfig.ConfigPath),
-		},
-		ImagePullPolicy: pullPolicy,
-		SecurityContext: &v1.SecurityContext{
-			// TODO: Determine SELINUX options needed to remove privileged
-			Privileged: &privTrue,
-		},
-	}
-
-	spec.InitContainers = append(spec.InitContainers, initContainer)
-
-	if p.EnableCoreDump {
-		spec.InitContainers = append(spec.InitContainers, enableCoreDumpContainer)
-	}
+// ProxySettings contains configurable settings for the sidecar mesh
+// in string form for proper template interpolation
+type ProxySettings struct {
+	DrainDuration          string
+	ParentShutdownDuration string
+	DiscoveryRefreshDelay  string
+	ConnectTimeout         string
+	ControlPlaneAuthPolicy string
+}
 
-	// sidecar proxy container
-	args := []string{"proxy", "sidecar"}
+func injectIntoSpec(p *Params, spec *v1.PodSpec, metadata *metav1.ObjectMeta) {
 
-	if p.Verbosity > 0 {
-		args = append(args, "-v", strconv.Itoa(p.Verbosity))
-	}
+	pSettings := ProxySettings{timeString(p.Mesh.DefaultConfig.DrainDuration),
+		timeString(p.Mesh.DefaultConfig.ParentShutdownDuration),
+		timeString(p.Mesh.DefaultConfig.DiscoveryRefreshDelay),
+		timeString(p.Mesh.DefaultConfig.ConnectTimeout), p.Mesh.DefaultConfig.ControlPlaneAuthPolicy.String()}
 
-	serviceCluster := p.Mesh.DefaultConfig.ServiceCluster
+	data := struct {
+		Spec           *v1.PodSpec
+		ServiceCluster string
+		MConfig        *Params
+		PConfig        *ProxySettings
+	}{spec, "", p, &pSettings}
```

Comment:
It would be nice to move towards defining the templated configuration scheme in terms of more publicly visible types. List of k8s containers/volumes is reasonable as is Mesh and Proxy config from istio/api. One-off types defined in inject.go is less discoverable. Can we try to fix this before the revised templates are in wide spread use?

Reply:
Yes do you think a SidecarTemplate type would suffice?

Reply:
Sure. I would try to avoid redefining new subtypes for global configuration that can be discovered elsewhere (e.g. ProxConfig proto). We could either encode that directly in the list of containers or fetch it from an existing source (e.g. ProxyConfig ConfigMap).